File: Problem2.py
'''

Time Complexity : O(n)
Space Complexity : O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach
Using BFS, I traverse thru the list level-wise and use two queues. One to track elements added level wise and the other queue
to store the corresponding parent element.

I use a conditional statement to check if both nodes are on the same level and don't have same parents
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        q = collections.deque()
        par = collections.deque()
        q.append(root)
        par.append(root.val)

        level = -1
        pos = [-1,-2,-3,-4]
        # [level1,parent1,level2,parent2]

        while q:
            level += 1

            for i in range(len(q)):
                node = q.popleft()
                parent = par.popleft()

                if node:
                    if node.val == x:
                        pos[0] = level
                        pos[1] = parent

                    if node.val == y:
                        pos[2] = level
                        pos[3] = parent

                    q.append(node.left)
                    par.append(node.val)
                    q.append(node.right)
                    par.append(node.val)

            if level > 1 and pos[0] == pos[2] and pos[1] != pos[3]:
                return True

        return False

File: problem-2.go
package main

/*time Complexity = O(n)
space Complexity = O(h)*/

func isCousins(root *TreeNode, x int, y int) bool {
	if root == nil {
		return false
	}

	var queue []*TreeNode
	queue = append(queue, root)

	for len(queue) > 0 {
		levelSize := len(queue)

		foundX, foundY := false, false

		for i := 0; i < levelSize; i++ {
			node := queue[0]
			queue = queue[1:]

			if node.Val == x {
				foundX = true
			}
			if node.Val == y {
				foundY = true
			}

			if node.Left != nil && node.Right != nil {
				if (node.Left.Val == x && node.Right.Val == y) || (node.Left.Val == y && node.Right.Val == x) {
					return false
				}
			}

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		if foundX && foundY {
			return true
		} else if foundX || foundY {
			return false
		}
	}

	return false
}
